aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -17,23 +17,212 @@
 #include <cmath>
 
 namespace at { namespace native {
+static inline Tensor _fft(
+    const Tensor &self, int64_t signal_ndim, bool complex_input,
+    const bool complex_output, bool inverse, IntArrayRef signal_sizes,
+    fft_norm_mode normalization, bool onesided);
+
+namespace {
+
+ScalarType promote_type(ScalarType type, bool require_complex) {
+  if (at::isComplexType(type)) {
+    return type;
+  }
+  // Promote integral to double
+  if (!at::isFloatingType(type)) {
+    return require_complex ? kComplexDouble : kDouble;
+  }
+  // Promote half to float
+  if (require_complex) {
+    return type == kDouble ? kComplexDouble : kComplexFloat;
+  } else {
+    return type == kDouble ? kDouble : kFloat;
+  }
+}
+
+Tensor promote_tensor(const Tensor& t, bool require_complex=false) {
+  auto cur_type = t.scalar_type();
+  auto new_type = promote_type(cur_type, require_complex);
+  return (cur_type == new_type) ? t : t.to(new_type);
+}
+
+fft_norm_mode norm_from_string(c10::optional<std::string> norm, bool forward) {
+  if (!norm || *norm == "backward") {
+    return forward ? fft_norm_mode::none : fft_norm_mode::by_n;
+  }
+
+  if (*norm == "forward") {
+    return forward ? fft_norm_mode::by_n : fft_norm_mode::none;
+  }
+
+  if (*norm == "ortho") {
+    return fft_norm_mode::by_root_n;
+  }
+
+  TORCH_CHECK(false, "Invalid normalization mode: \"", *norm, "\"")
+}
+
+Tensor fix_shape(Tensor x, IntArrayRef dims, IntArrayRef sizes) {
```

Comment:
Better add a comment explaining this function. Would something like `pad_fft` be a more descriptive name?

Reply:
`pad` functions usually takes a padding amount relative to the current size, whereas this takes the desired shape (in the fft dimensions).

Reply:
Good point. Maybe something more like "resize_fft_input"?

Reply:
Renamed to `resize_fft_input`.